gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/taskexecution/IllegalMethodCalledDuringTaskExecution.java

```diff
@@ -0,0 +1,104 @@
+        required for the task's functionality. For example, instead getProject().version(), you should declare the
+        project version as an `@Input public abstract Property<String>`.
+
+        Doing so improves performance in two ways:
+        1. It makes your tasks compatible with the configuration cache
+        2. It increases task parallelism. When two tasks, such as printProjectName and printProjectVersion, both
+        require the same Project object as input, they cannot run in parallel due to prevent concurrent access.
+        However, if their inputs are changed to Provider<String> name and Provider<String> version respectively,
+        the tasks become independent and can execute in parallel.
+        """)
+public final class IllegalMethodCalledDuringTaskExecution extends GradleGuideBugChecker
+        implements BugChecker.MethodInvocationTreeMatcher {
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+        Optional<Violation> firstMatchingViolation = violations.stream()
+                .filter(violation -> violation.matches(tree, state))
+                .findFirst();
+        if (firstMatchingViolation.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+
+        CompilationUnitTree compilationUnit = state.getPath().getCompilationUnit();
+        MethodCallGraph callGraph = MethodCallGraph.getOrBuild(compilationUnit);
+
+        Optional<MethodTree> enclosingMethodMaybe = Optional.ofNullable(ASTHelpers.findEnclosingMethod(state));
+        if (enclosingMethodMaybe.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+
+        MethodSymbol enclosingMethod = ASTHelpers.getSymbol(enclosingMethodMaybe.get());
+        // Find transitive callers of enclosing method first, as the illegal method e.g. getProject() is defined in
+        // Gradle source, not the source file this BugChecker is running on.
+        // If we included methods defined externally into the call graph,
+        // IllegalMethodCalledDuringTaskExecutionTest#getProject_in_constructor_ok will fail.
+        Set<MethodSymbol> transitiveCallersOfIllegalMethod = callGraph.transitiveCallers(enclosingMethod);
+
+        boolean isInvokedAtTaskExecution = StreamEx.of(transitiveCallersOfIllegalMethod)
+                .append(enclosingMethod)
+                .anyMatch(caller -> {
+                    MethodTree callerTree = ASTHelpers.findMethod(caller, state);
+                    return Tasks.isTaskAction(callerTree, state) || Tasks.overridesExecute(callerTree, state);
+                });
+        if (isInvokedAtTaskExecution) {
+            firstMatchingViolation.get().fixOrReport(tree, state);
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    // Only the first matching violation is applied. Put more specific violations first.
+    private final List<Violation> violations = List.of(new GetProjectGetLogger(this), new GetProject(this));
```

Comment:
Let's put this at the top of the class - it's unusual to have fields anywhere other than the top of the class.